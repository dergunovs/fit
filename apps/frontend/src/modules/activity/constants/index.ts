import { TActivityChartType } from 'fitness-tracker-contracts';

export const URL_ACTIVITY = '/activities';
export const URL_ACTIVITY_CREATE = `${URL_ACTIVITY}/create`;
export const URL_ACTIVITY_EDIT = `${URL_ACTIVITY}/edit`;

export const URL_ACTIVITY_ADMIN = '/activities_admin';
export const URL_ACTIVITY_ADMIN_EDIT = `${URL_ACTIVITY_ADMIN}/edit`;

export const ACTIVITY_LIST_HEADERS = [{ title: 'Создано' }, { title: 'Пользователь' }];

export const CHART_TYPES: { title: string; value: TActivityChartType }[] = [
  { title: 'Занятия', value: 'activity' },
  { title: 'Подходы', value: 'set' },
  { title: 'Повторы', value: 'repeat' },
  { title: 'Мышцы', value: 'muscle' },
];

export const ACTIVITY_STATISTICS_GAP = 30;
