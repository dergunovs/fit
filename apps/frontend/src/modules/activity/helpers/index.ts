import { IActivity, IExerciseDone, IExerciseStatistics } from 'fitness-tracker-contracts';
import { createTempId, formatDate, formatDuration, subtractDates } from 'mhz-helpers';
import { toast } from 'mhz-ui';

export function getPotentialActivityDuration(
  exercises: IExerciseDone[],
  exerciseStatistics: IExerciseStatistics[],
  averageRestPercent?: number
) {
  if (!averageRestPercent) return '-';

  const totalDuration = exercises.reduce((acc, exercise) => {
    const averageDuration =
      exerciseStatistics.find((choosenExericse) => choosenExericse._id === exercise.exercise?._id)?.averageDuration ||
      0;

    return acc + averageDuration * exercise.repeats;
  }, 0);

  const durationWithRest = Math.round(totalDuration / (1 - averageRestPercent / 100));

  return formatDuration(durationWithRest);
}

export function convertActivityCalendarEvents(activities?: IActivity[]) {
  return activities?.map((activity: IActivity) => {
    return {
      _id: activity._id,
      start: new Date(`${activity.dateCreated}`),
      end: new Date(`${activity.dateUpdated}`),
      title: '+',
      content: activity.exercises,
    };
  });
}

export function generateActivityExercises(exercisesDone: IExerciseDone[]): IExerciseDone[] {
  const activityExercises = exercisesDone.map((exercise) => {
    return {
      _id: createTempId(),
      exercise: {
        _id: exercise.exercise?._id || '',
        title: exercise.exercise?.title || '',
        muscleGroups: exercise.exercise?.muscleGroups || [],
        isWeights: exercise.exercise?.isWeights || false,
        isWeightsRequired: exercise.exercise?.isWeightsRequired || false,
      },
      repeats: exercise.repeats,
      weight: exercise.weight,
    };
  });

  return activityExercises?.length ? [...activityExercises] : [];
}

export function getToFailurePercent(exercises: IExerciseDone[]) {
  const allExercises = exercises.length;
  const toFailureExercises = exercises.filter((exercise) => exercise.isToFailure).length;

  return `${Math.floor((toFailureExercises / allExercises) * 100)}%`;
}

export function getRestPercent(exercises: IExerciseDone[], start?: Date | null | string, end?: Date | null | string) {
  const activityDuration = Number(subtractDates(end, start, true));
  const exercisesDuration = exercises.reduce((acc, current) => acc + (current.duration || 0), 0);

  return `${Math.floor(100 - (exercisesDuration / activityDuration) * 100)}%`;
}

export async function copyActivityToClipboard(
  exercises: IExerciseDone[],
  start?: Date | null | string,
  end?: Date | null | string
) {
  const textHeader = `${formatDate(start, 'ru')}, длительность: ${subtractDates(end, start)}
Подходы: ${exercises.length}, отказы: ${getToFailurePercent(exercises)}, отдых: ${getRestPercent(exercises, start, end)}.

${exercises
  .map((exercise, index) => {
    return `${index + 1}. ${exercise.exercise?.title} x${exercise.repeats} ${exercise.weight ? `${exercise.weight}кг` : ''} ${formatDuration(exercise.duration)} ${exercise.isToFailure ? 'ДО ОТКАЗА' : ''}\n`;
  })
  .join('')}`;

  await navigator.clipboard.writeText(textHeader);

  toast.success('Скопировано в буфер');
}
