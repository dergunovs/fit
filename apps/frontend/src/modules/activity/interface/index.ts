export interface IActivityCalendarEvent<T> {
  _id?: string;
  start: Date | null;
  end: Date | null;
  title: string;
  content: T[];
  color?: string;
}

export interface ICalendarDates {
  dateFrom: string;
  dateTo: string;
}

export interface ITimelineStep {
  left: number;
  right: number;
  type: 'rest' | 'exercise';
}
