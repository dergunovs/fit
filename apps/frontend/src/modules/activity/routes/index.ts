import {
  URL_ACTIVITY_CREATE,
  URL_ACTIVITY_EDIT,
  URL_ACTIVITY_ADMIN,
  URL_ACTIVITY_ADMIN_EDIT,
} from '@/activity/constants';

export const activityRoutes = [
  {
    path: URL_ACTIVITY_CREATE,
    name: 'ActivityCreate',
    component: () => import('@/activity/pages/ActivityCreatePage.vue'),
  },
  {
    path: `${URL_ACTIVITY_EDIT}/:activity`,
    name: 'ActivityPassing',
    component: () => import('@/activity/pages/ActivityPassingPage.vue'),
  },
  {
    path: URL_ACTIVITY_ADMIN,
    name: 'ActivityList',
    component: () => import('@/activity/pages/ActivityListPage.vue'),
    meta: { layout: 'admin' },
  },
  {
    path: `${URL_ACTIVITY_ADMIN_EDIT}/:activity`,
    name: 'ActivityEdit',
    component: () => import('@/activity/pages/ActivityEditPage.vue'),
    meta: { layout: 'admin' },
  },
];
