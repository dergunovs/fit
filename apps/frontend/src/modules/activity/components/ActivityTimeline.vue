<template>
  <div :class="$style.timeline">
    <div
      v-for="(step, index) in generateTimeline(props.exercises, props.start, ratio)"
      :key="index"
      :class="$style.step"
      :data-type="step.type"
      :style="{ left: `${step.left}px`, right: `${step.right}px`, width: `${step.right - step.left}px` }"
      data-test="activity-timeline-step"
    ></div>
  </div>
</template>

<script setup lang="ts">
import { computed } from 'vue';
import { IExerciseDone } from 'fitness-tracker-contracts';

import { generateTimeline } from '@/exercise/helpers';

interface IProps {
  exercises: IExerciseDone[];
  start: Date | string | null;
}

const props = defineProps<IProps>();

const width = 308;

const lastStepDate = props.exercises[props.exercises.length - 1].dateUpdated;

const ratio =
  lastStepDate && props.start ? (new Date(lastStepDate).getTime() - new Date(props.start).getTime()) / width : 0;

const widthWithPx = computed(() => `${width}px`);
</script>

<style module lang="scss">
.timeline {
  position: relative;
  width: v-bind(widthWithPx);
  height: 40px;
  margin-bottom: 8px;
  border-radius: 8px;
}

.step {
  position: absolute;
  height: 40px;

  &[data-type='rest'] {
    background-color: var(--color-gray);
  }

  &[data-type='exercise'] {
    background-color: var(--color-primary);
  }

  &:first-child {
    border-top-left-radius: 8px;
    border-bottom-left-radius: 8px;
  }

  &:last-child {
    border-top-right-radius: 8px;
    border-bottom-right-radius: 8px;
  }
}
</style>
