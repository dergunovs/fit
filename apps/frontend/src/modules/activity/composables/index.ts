import { ref } from 'vue';

import { ICalendarDates } from '@/activity/interface';

export function useActivityCalendar() {
  const dateFrom = ref('');
  const dateTo = ref('');

  const isDatesReady = ref(false);

  function updateDates(dates: ICalendarDates) {
    isDatesReady.value = true;

    dateFrom.value = dates.dateFrom;
    dateTo.value = dates.dateTo;
  }

  return { dateFrom, dateTo, isDatesReady, updateDates };
}
