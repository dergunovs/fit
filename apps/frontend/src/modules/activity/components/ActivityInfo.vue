<template>
  <div :class="$style.info" :data-scrollable="props.isPopup" data-test="activity-info">
    <UiFlex gap="4" align="center" wrap>
      <ActivityTimeline
        v-if="props.isPopup && props.exercises[0].dateUpdated"
        :exercises="props.exercises"
        :start="props.start"
        data-test="activity-timeline"
      />

      <span data-test="activity-info-start">
        <IconDate width="16" height="16" /> {{ formatDate(props.start, 'ru') }}
      </span>

      <span v-if="isExercisesDone" data-test="activity-info-duration">
        <IconDuration width="16" height="16" /> {{ subtractDates(props.end, props.start) }}
      </span>

      <span>
        Подходы:
        <span data-test="activity-info-sets">{{ props.exercises.length }}</span
        ><template v-if="isExercisesDone"
          >, отказы: <span data-test="activity-info-to-failure-percent">{{ getToFailurePercent(props.exercises) }}</span
          >, отдых:
          <span data-test="activity-info-rest-percent">
            {{ getRestPercent(props.exercises, props.start, props.end) }}
          </span>
        </template>
      </span>

      <UiButton
        v-if="isExercisesDone"
        @click="copyActivityToClipboard(props.exercises, props.start, props.end)"
        layout="plain"
        data-test="activity-info-copy"
      >
        Копировать
      </UiButton>
    </UiFlex>

    <ExerciseMuscleGroupStatistics :exercises="props.exercises" data-test="activity-info-muscle-groups" />

    <UiFlex column>
      <ExerciseTitle
        v-for="(exercise, index) in props.exercises"
        :key="exercise._id"
        :exercise="exercise"
        :isHideTitle="isPrevExerciseSame(props.exercises, index, exercise.exercise?._id)"
        :isHideDuration="isFutureActivity"
        data-test="activity-info-exercise"
      />
    </UiFlex>

    <UiFlex justify="space-between">
      <UiButton
        v-if="isAuth && isPopup && !isFutureActivity"
        @click="router.push(`${URL_ACTIVITY_CREATE}?copy=${props.id}`)"
        data-test="activity-info-repeat"
      >
        Скопировать
      </UiButton>

      <UiButton
        v-if="isAuth && isPopup && isFutureActivity"
        @click="router.push(`${URL_ACTIVITY_EDIT}/${props.id}`)"
        data-test="activity-info-start"
      >
        Начать
      </UiButton>

      <UiButton
        v-if="isAuth && isPopup"
        @click="isShowConfirm = true"
        layout="secondary"
        data-test="activity-info-start"
      >
        Удалить
      </UiButton>
    </UiFlex>

    <UiModal v-model="isShowConfirm" isConfirm @confirm="mutateDelete(id)" width="360" data-test="activity-info-modal">
      Подтверждаете удаление?
    </UiModal>
  </div>
</template>

<script setup lang="ts">
import { ref, computed } from 'vue';
import { useRouter } from 'vue-router';
import { API_ACTIVITY, API_ACTIVITY_CALENDAR, IExerciseDone } from 'fitness-tracker-contracts';
import { toast, UiButton, UiFlex, UiModal } from 'mhz-ui';
import { formatDate, subtractDates, isAuth, useQueryClient } from 'mhz-helpers';

import ActivityTimeline from '@/activity/components/ActivityTimeline.vue';
import ExerciseTitle from '@/exercise/components/ExerciseTitle.vue';
import ExerciseMuscleGroupStatistics from '@/exercise/components/ExerciseMuscleGroupStatistics.vue';
import IconDate from '@/common/icons/date.svg';
import IconDuration from '@/common/icons/duration.svg';

import { URL_ACTIVITY_CREATE, URL_ACTIVITY_EDIT } from '@/activity/constants';
import { copyActivityToClipboard, getRestPercent, getToFailurePercent } from '@/activity/helpers';
import { isPrevExerciseSame } from '@/exercise/helpers';
import { activityService } from '@/activity/services';

interface IProps {
  id: string;
  start: Date | null | string;
  end: Date | null | string;
  exercises: IExerciseDone[];
  isPopup?: boolean;
}

const props = defineProps<IProps>();
const emit = defineEmits<{ delete: [] }>();

const router = useRouter();
const queryClient = useQueryClient();

const isShowConfirm = ref(false);

const isFutureActivity = computed(() => !!(props.start && props.start > new Date()));
const isExercisesDone = computed(() => props.exercises.some((exercise) => exercise.isDone));

const { mutate: mutateDelete } = activityService.delete({
  onSuccess: async () => {
    queryClient.removeQueries({ queryKey: [API_ACTIVITY, API_ACTIVITY_CALENDAR] });
    await queryClient.refetchQueries({ queryKey: [API_ACTIVITY, API_ACTIVITY_CALENDAR] });
    toast.success('Занятие удалено');
    emit('delete');
  },
});
</script>

<style module lang="scss">
.info {
  display: flex;
  flex-direction: column;
  gap: 16px;

  &[data-scrollable='true'] {
    max-height: 64dvh;
    overflow-y: auto;
  }
}
</style>
