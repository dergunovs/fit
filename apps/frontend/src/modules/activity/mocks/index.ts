import { ref, computed } from 'vue';
import { vi } from 'vitest';
import { TDeleteActivityDTO, TGetActivitiesDTO } from 'fitness-tracker-contracts';

import * as activityComposables from '@/activity/composables';
import { activityService } from '@/activity/services';
import {
  ACTIVITIES_CALENDAR_FIXTURE,
  ACTIVITIES_FIXTURE,
  ACTIVITIES_STATISTICS_FIXTURE,
  ACTIVITY_FIXTURE,
  ACTIVITIES_CHART_FIXTURE,
} from '@/activity/fixtures';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { IOnSuccess } from '@/common/interface';

const dateFrom = ref('2024-11-24T21:00:00.000Z');
const dateTo = ref('2025-01-05T20:59:59.000Z');
const isDatesReady = ref(true);
const spyUpdateDates = vi.fn();

const spyGetActivity = vi.spyOn(activityService, 'getOne').mockReturnValue(mockQueryReply(ACTIVITY_FIXTURE));

const getActivitiesData: TGetActivitiesDTO = { data: ACTIVITIES_FIXTURE, total: ACTIVITIES_FIXTURE.length };

const spyGetActivities = vi
  .spyOn(activityService, 'getMany')
  .mockImplementation(() => mockQueryReply(getActivitiesData));

const spyDeleteActivity = vi.fn();

const onSuccess: IOnSuccess = {
  delete: undefined,
};

vi.spyOn(activityService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) onSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteActivityDTO, string>(spyDeleteActivity);
});

const spyUseActivityCalendar = vi.spyOn(activityComposables, 'useActivityCalendar').mockReturnValue({
  dateFrom,
  dateTo,
  isDatesReady,
  updateDates: spyUpdateDates,
});

const computedActivityCalendarEvents = computed(() =>
  activityComposables.convertActivityCalendarEvents(ACTIVITIES_CALENDAR_FIXTURE)
);

const spyComputedActivityCalendarEvents = vi
  .spyOn(activityComposables, 'computedActivityCalendarEvents')
  .mockImplementation(() => computedActivityCalendarEvents);

const spyGetActivitiesCalendar = vi
  .spyOn(activityService, 'getCalendar')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_CALENDAR_FIXTURE));

const spyGetActivitiesStatistics = vi
  .spyOn(activityService, 'getStatistics')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_STATISTICS_FIXTURE));

const spyGetActivitiesChart = vi
  .spyOn(activityService, 'getChart')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_CHART_FIXTURE));

export {
  spyGetActivities,
  spyGetActivity,
  spyGetActivitiesStatistics,
  spyGetActivitiesChart,
  spyDeleteActivity,
  spyUseActivityCalendar,
  spyComputedActivityCalendarEvents,
  spyGetActivitiesCalendar,
  spyUpdateDates,
  dateFrom,
  dateTo,
  isDatesReady,
  computedActivityCalendarEvents,
  getActivitiesData,
  onSuccess,
};
