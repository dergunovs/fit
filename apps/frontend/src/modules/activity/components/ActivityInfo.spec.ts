import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { formatDate, subtractDates, setAuth, dataTest } from 'mhz-helpers';
import { UiModal } from 'mhz-ui';
import { API_ACTIVITY, API_ACTIVITY_CALENDAR } from 'fitness-tracker-contracts';

import ActivityInfo from './ActivityInfo.vue';
import ActivityTimeline from './ActivityTimeline.vue';
import MuscleStatistics from '@/muscle/components/MuscleStatistics.vue';
import ExerciseTitle from '@/exercise/components/ExerciseTitle.vue';

import { wrapperFactory } from '@/common/test';
import { EXERCISES_DONE_FIXTURE } from '@/exercise/fixtures';
import { getRestPercent, getToFailurePercent } from '@/activity/helpers';
import { mockOnSuccess, spyDeleteActivity } from '@/activity/mocks';
import { URL_ACTIVITY_CREATE, URL_ACTIVITY_EDIT } from '@/activity/constants';
import { MUSCLES_FIXTURE } from '@/muscle/fixtures';
import { generateMuscleStatistics } from '@/muscle/helpers';
import { spyGetMuscles } from '@/muscle/mocks';
import {
  spyRefetchQueries,
  spyRemoveQueries,
  spyRouterPush,
  spyToastSuccess,
  spyCopyToClipboard,
} from '@/common/mocks';

const activityInfo = dataTest('activity-info');
const timeline = dataTest('activity-timeline');
const startDate = dataTest('activity-info-start-date');
const durationTime = dataTest('activity-info-duration');
const sets = dataTest('activity-info-sets');
const toFailurePercent = dataTest('activity-info-to-failure-percent');
const restPercent = dataTest('activity-info-rest-percent');
const muscleStatistics = dataTest('activity-info-muscle-statistics');
const exercise = dataTest('activity-info-exercise');
const copyActivity = dataTest('activity-info-copy');
const startActivity = dataTest('activity-info-start');
const deleteActivity = dataTest('activity-info-delete');
const copyActivityToClipboard = dataTest('activity-info-copy-to-clipboard');
const modal = dataTest('activity-info-modal');

const id = '123';
const start = new Date('01-01-2025');
const end = new Date('01-02-2025');
const exercises = EXERCISES_DONE_FIXTURE;
const isPopup = false;

let wrapper: VueWrapper<InstanceType<typeof ActivityInfo>>;

beforeEach(() => {
  wrapper = wrapperFactory(ActivityInfo, { id, start, end, exercises, isPopup });
});

enableAutoUnmount(afterEach);

describe('ActivityInfo', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ActivityInfo)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('info is scrollable in popup', async () => {
    expect(wrapper.find(activityInfo).attributes('data-scrollable')).toBe(isPopup.toString());

    await wrapper.setProps({ isPopup: true });

    expect(wrapper.find(activityInfo).attributes('data-scrollable')).toBe(true.toString());
  });

  it('shows timeline in popup', async () => {
    expect(wrapper.find(timeline).exists()).toBe(false);

    await wrapper.setProps({ isPopup: true });

    expect(wrapper.findComponent<typeof ActivityTimeline>(timeline).vm.$props.exercises).toStrictEqual(exercises);
    expect(wrapper.findComponent<typeof ActivityTimeline>(timeline).vm.$props.start).toStrictEqual(start);
  });

  it('shows start and duration time', async () => {
    expect(wrapper.find(startDate).text()).toBe(formatDate(start, 'ru'));
    expect(wrapper.find(durationTime).text()).toBe(subtractDates(end, start, 'ru'));
  });

  it('shows sets, to failure and rest percents', async () => {
    expect(wrapper.find(sets).text()).toBe(exercises.length.toString());
    expect(wrapper.find(toFailurePercent).text()).toBe(getToFailurePercent(exercises));
    expect(wrapper.find(restPercent).text()).toBe(getRestPercent(exercises, 'ru', start, end));
  });

  it('gets and sets data to muscle statistics table', async () => {
    expect(spyGetMuscles).toBeCalledTimes(1);

    const statistics = generateMuscleStatistics(exercises, MUSCLES_FIXTURE, 'ru');

    expect(wrapper.findComponent<typeof MuscleStatistics>(muscleStatistics).vm.$props.statistics).toStrictEqual(
      statistics
    );
  });

  it('shows exercises', async () => {
    expect(wrapper.findAll(exercise).length).toBe(exercises.length);
    expect(wrapper.findAllComponents<typeof ExerciseTitle>(exercise)[0].vm.$props.exercise).toStrictEqual(exercises[0]);
  });

  it('copies activity in popup mode', async () => {
    expect(wrapper.find(copyActivity).exists()).toBe(false);
    expect(spyRouterPush).toBeCalledTimes(0);

    setAuth(true);
    await wrapper.setProps({ isPopup: true });

    expect(wrapper.find(copyActivity).exists()).toBe(true);

    await wrapper.find(copyActivity).trigger('click');

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(`${URL_ACTIVITY_CREATE}?copy=${id}`);
  });

  it('starts activity in popup', async () => {
    expect(wrapper.find(startActivity).exists()).toBe(false);

    setAuth(true);
    await wrapper.setProps({ isPopup: true, start: new Date('01-01-3000') });

    expect(wrapper.find(startActivity).exists()).toBe(true);

    await wrapper.find(startActivity).trigger('click');

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(`${URL_ACTIVITY_EDIT}/${id}`);
  });

  it('deletes activity in popup', async () => {
    expect(spyDeleteActivity).toBeCalledTimes(0);
    expect(spyRemoveQueries).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);

    expect(wrapper.emitted()).not.toHaveProperty('delete');

    expect(wrapper.find(deleteActivity).exists()).toBe(false);

    setAuth(true);
    await wrapper.setProps({ isPopup: true });

    expect(wrapper.find(deleteActivity).exists()).toBe(true);

    expect(wrapper.find(modal).attributes('modelvalue')).toBe('false');

    await wrapper.find(deleteActivity).trigger('click');

    expect(wrapper.find(modal).attributes('modelvalue')).toBe('true');

    wrapper.findComponent<typeof UiModal>(modal).vm.$emit('confirm');

    expect(spyDeleteActivity).toBeCalledTimes(1);
    expect(spyDeleteActivity).toBeCalledWith(id);

    await mockOnSuccess.delete?.();

    expect(spyRemoveQueries).toBeCalledTimes(1);
    expect(spyRemoveQueries).toBeCalledWith({ queryKey: [API_ACTIVITY, API_ACTIVITY_CALENDAR] });

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_ACTIVITY, API_ACTIVITY_CALENDAR] });

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(wrapper.emitted('delete')).toHaveLength(1);
  });

  it('copies activity to clipboard', async () => {
    expect(spyCopyToClipboard).toBeCalledTimes(0);

    await wrapper.find(copyActivityToClipboard).trigger('click');

    expect(spyCopyToClipboard).toBeCalledTimes(1);
  });
});
