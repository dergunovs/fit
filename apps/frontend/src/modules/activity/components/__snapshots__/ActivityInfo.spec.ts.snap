// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActivityInfo > matches snapshot 1`] = `
"<div class="_info_d22753" data-scrollable="false" data-test="activity-info">
  <div gap="4" align="center" wrap="">
    <!--v-if-->
    <div>
      <div gap="2" align="center" data-test="activity-info-start-date">
        <anonymous-stub width="16" height="16"></anonymous-stub> 1 янв. 2025 г.
      </div>
      <div gap="2" align="center" data-test="activity-info-duration">
        <anonymous-stub width="16" height="16"></anonymous-stub> 1440 мин. 00 сек.
      </div>
    </div>
    <div>Подходы: <span data-test="activity-info-sets">2</span>, отказы: <span data-test="activity-info-to-failure-percent">50%</span>, отдых: <span data-test="activity-info-rest-percent">99%</span></div><button layout="plain" data-test="activity-info-copy-to-clipboard">Скопировать в буфер</button>
  </div>
  <muscle-statistics-stub statistics="[object Object],[object Object],[object Object],[object Object]" data-test="activity-info-muscle-statistics"></muscle-statistics-stub>
  <div column="">
    <exercise-title-stub exercise="[object Object]" ishidetitle="false" ishideduration="false" data-test="activity-info-exercise"></exercise-title-stub>
    <exercise-title-stub exercise="[object Object]" ishidetitle="false" ishideduration="false" data-test="activity-info-exercise"></exercise-title-stub>
  </div>
  <div justify="space-between"></div>
  <div modelvalue="false" isconfirm="" width="360" lang="ru" data-test="activity-info-modal">Подтверждаете удаление? </div>
</div>"
`;
